File: leetCode/c/p2_AddTwoNumbers/lc2.c
/* You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order, 
and each of their nodes contains a single digit. Add the two numbers and return the sum as a linked list.
You may assume the two numbers do not contain any leading zero, except the number 0 itself. */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>

/**
 * Definition for singly-linked list.
*/
struct ListNode {
    int val;
    struct ListNode *next;
};

void listInitialize(struct ListNode **list) {
   
    *list = (struct ListNode*) malloc(sizeof(*list)); 

    (*list)->val = -1;
    (*list)->next = NULL;
}

void listInsertArray(struct ListNode *list, int *array, int arraySize) {
    
    int firstIndex = 0;
     
    /* Check if first element */
    if(list->val < 0) {
        list->val = array[0];
        firstIndex = 1;
        if(arraySize == 1) return;
    }

    /* Go to the end of linked list */ 
    while(list->next != NULL) {
        list = list->next;
    }

    for(int i = firstIndex; i<arraySize; i++) {      
        struct ListNode *temp = (struct ListNode*) malloc(sizeof(*temp));
        temp->val = array[i];
        list->next = temp; 
        list = list->next;
    }    
    
    list->next = NULL;
}

void listInsertSingle(struct ListNode *list, int val) {
    
    struct ListNode *temp = (struct ListNode*) malloc(sizeof(*temp));  
    
    /* Check if first element */ 
    if(list->val < 0) {
        list->val = val;
        return;
    }
 
    /* Go to the end of linked list */ 
    while(list->next != NULL) {
        list = list->next;
    }

    temp->val = val;
    temp->next = NULL;
    list->next = temp;
}

void listPrint(struct ListNode *list) {
    
    struct ListNode *curr = list;
 
    while(curr->next != NULL) {
        printf("%d ", curr->val); 
        curr = curr->next;
    }

    printf("%d\n", curr->val);
}

void listFree(struct ListNode *list) {
    
    struct ListNode *tmp;

    while(list->next != NULL) {
        tmp = list;
        list = list->next;
        free(tmp);
    }
}

int listFormNumber(struct ListNode *list) {
    
    int sum = 0, mul = 1;

    do { 
        sum += (list->val)*mul;
        mul *= 10;
        list = list->next;
    } while(list != NULL);

    return sum;
}

struct ListNode* addTwoNumbers(struct ListNode *l1, struct ListNode *l2) {
   
    struct ListNode *tmp;
    int sum = 0, mul = 1; 

    listInitialize(&tmp);

    sum += listFormNumber(l1);
    sum += listFormNumber(l2);
      
    while(true) {
        listInsertSingle(tmp, sum % 10);
        sum /= 10;
        if(sum == 0)
            break;
    }

    return tmp;
}

/* Test */

void testFunction(struct ListNode *l1, struct ListNode *l2, struct ListNode *sol, int testNumber) {
      
    bool identical = true; 
    float timeSpent = 0;
    struct ListNode *curr; 
    
    clock_t tic = clock();
    
    curr = addTwoNumbers(l1, l2);
    
    clock_t toc = clock();
 
    timeSpent = (double) (toc - tic) * 1000000 / CLOCKS_PER_SEC;
    
    if(curr->next == NULL && sol->next == NULL) {
        if(curr->val != sol->val)
            identical = false;
    } 

    do {
        if(sol->next == NULL && curr -> next != NULL) {
            identical = false;
            break;
        } 
        if(curr->val != sol->val) {
            identical = false;
            break;
        }
        curr = curr->next;
        sol = sol->next;  
    } while(curr != NULL);

    printf("\nTEST%d: ", testNumber);

    if(identical) 
        printf("PASSED IN %.2fus\n", timeSpent);
    else  
        printf("FAILED\n");
}

int main(void) {
    
    struct ListNode *l1, *l2, *sol; 

    /* Test 1 */

    listInitialize(&l1);
    listInitialize(&l2);
    listInitialize(&sol);
    
    int array1[3] = {2, 4, 3};
    listInsertArray(l1, array1, 3);
    
    int array2[3] = {5, 6, 4};
    listInsertArray(l2, array2, 3);

    int array3[3] = {7, 0, 8};
    listInsertArray(sol, array3, 3);

    testFunction(l1, l2, sol, 1);
         
    listFree(l1);
    listFree(l2);
    listFree(sol);
    
    /* Test 2 */
 
    listInitialize(&l1);
    listInitialize(&l2);
    listInitialize(&sol);
 
    listInsertSingle(l1, 0);

    listInsertSingle(l2, 0); 

    listInsertSingle(sol, 0);

    testFunction(l1, l2, sol, 2);
    
    listFree(l1);
    listFree(l2);
    listFree(sol);
    
    /* Test 3 */
         
    listInitialize(&l1);
    listInitialize(&l2);
    listInitialize(&sol);
    
    int array4[7] = {9, 9, 9, 9, 9, 9, 9}; 
    listInsertArray(l1, array4, 7);

    int array5[4] = {9, 9, 9, 9};
    listInsertArray(l2, array5, 4);
    
    int array6[8] = {8, 9, 9, 9, 0, 0, 0, 1};
    listInsertArray(sol, array6, 8);

    testFunction(l1, l2, sol, 3);

    return EXIT_SUCCESS;
}

File: leetCode/python/p2_AddTwoNumbers/lc2.py
import time

# Definition of a singly-linked list.
class ListNode:
    def __init__(self, val = 0, next = None):
        self.val = val
        self.next = next

# Linked List class with a single head node.
class LinkedList:
    def __init__(self):
        self.head = None

    def insert_node(self, val):
        new_node = ListNode(val)
        # check if head already exists
        if(self.head):
            current = self.head 
            # go to the end of linked list
            while(current.next):
                current = current.next
            current.next = new_node
        else:
            self.head = new_node

    def insert_node_from_list(self, val: list):
        for i in range(len(val)):
            self.insert_node(val[i]) 

    def printLinkedList(self):
        current = self.head
        while(current):
            print(current.val, end = ' ')
            current = current.next
        print()

class Solution:
    """ 
    You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order, 
    and each of their nodes contains a single digit. Add the two numbers and return the sum as a linked list.
    You may assume the two numbers do not contain any leading zero, except the number 0 itself.
    """
    def __init__(self, l1: LinkedList, l2: LinkedList):
        self.l1 = l1
        self.l2 = l2
        self.sol = LinkedList()
        self.sol_num = 0
    
    def form_number_from_list(self, listIndex: int):
        if(listIndex == 2):
            current = self.l2.head
        else:
            current = self.l1.head
        mul = 1
        tmp = 0 
        while(current):
            tmp += current.val * mul
            mul *= 10
            current = current.next 
        return tmp
         
    def add_two_numbers(self): # changed from leetCode version 
        tmp = self.form_number_from_list(1) + self.form_number_from_list(2)
        self.sol_num = tmp
        while(tmp != 0):
            self.sol.insert_node(tmp % 10)
            tmp //= 10
        self.sol 

    def test_function(self, sol_num: int, test_num: int):
        start_time = time.time()
        self.add_two_numbers()                
        end_time = time.time()
        if(self.sol_num == sol_num):
            print('\nTEST' + str(test_num) + ': PASSED IN ' + '{:.2f}us'.format((end_time - start_time) * 1000000))
        else:
            print('\nTEST FAILED')
 
# Test 1

l1 = LinkedList()
tmp_list = [2, 4, 3]
l1.insert_node_from_list(tmp_list)

l2 = LinkedList()
tmp_list = [5, 6, 4] 
l2.insert_node_from_list(tmp_list)

s = Solution(l1, l2)
s.test_function(807, 1)

# Test 2

l1 = LinkedList()
l1.insert_node(0)

l2 = LinkedList()
l2.insert_node(0)

s = Solution(l1, l2)
s.test_function(0, 2)

# Test 3
l1 = LinkedList()
tmp_list = [9, 9, 9, 9, 9, 9, 9]
l1.insert_node_from_list(tmp_list)

l2 = LinkedList()
tmp_list = [9, 9, 9, 9]
l2.insert_node_from_list(tmp_list)

s = Solution(l1, l2)
s.test_function(10009998, 3)
